numpy/distutils/ccompiler_opt.py

```diff
@@ -0,0 +1,2358 @@
+                partial["VSX3"]["flags"] = "-mpower9-vector"
+
+            return partial
+
+        if self.cc_on_aarch64 and is_unix: return dict(
+            NEON = dict(
+                implies="NEON_FP16 NEON_VFPV4 ASIMD", autovec=True
+            ),
+            NEON_FP16 = dict(
+                implies="NEON NEON_VFPV4 ASIMD", autovec=True
+            ),
+            NEON_VFPV4 = dict(
+                implies="NEON NEON_FP16 ASIMD", autovec=True
+            ),
+            ASIMD = dict(
+                implies="NEON NEON_FP16 NEON_VFPV4", autovec=True
+            ),
+            ASIMDHP = dict(
+                flags="-march=armv8.2-a+fp16"
+            ),
+            ASIMDDP = dict(
+                flags="-march=armv8.2-a+dotprod"
+            ),
+            ASIMDFHM = dict(
+                flags="-march=armv8.2-a+fp16fml"
+            ),
+        )
+        if self.cc_on_armhf and is_unix: return dict(
+            NEON = dict(
+                # Avoid 'hard' since VPF doesn't have queue instruction like NEON
+                flags="-mfloat-abi=softfp -march=armv7-a+neon"
+            ),
+            NEON_FP16 = dict(
+                flags="-march=armv7-a+neon-fp16 -mfp16-format=ieee"
+            ),
+            NEON_VFPV4 = dict(
+                flags="-march=armv7-a+neon-vfpv4",
+            ),
+            ASIMD = dict(
+                flags="-march=armv8-a+simd",
+            ),
+            ASIMDHP = dict(
+                flags="-march=armv8.2-a+fp16"
+            ),
+            ASIMDDP = dict(
+                flags="-march=armv8.2-a+dotprod",
+            ),
+            ASIMDFHM = dict(
+                flags="-march=armv8.2-a+fp16fml"
+            )
+        )
+        # TODO: ARM MSVC
+        return {}
+
+    def __init__(self):
+        if self.conf_tmp_path is None:
+            import tempfile, shutil
+            tmp = tempfile.mkdtemp()
+            def rm_temp():
+                try:
+                    shutil.rmtree(tmp)
+                except:
+                    pass
+            atexit.register(rm_temp)
+            self.conf_tmp_path = tmp
+
+        if self.conf_cache_factors is None:
+            self.conf_cache_factors = [
+                os.path.getmtime(__file__),
+                self.conf_nocache
+            ]
+
+class _Distutils:
+    """A helper class that provides a collection of fundamental methods
+    implemented in a top of Python and NumPy Distutils.
+
+    The idea behind this class is to gather all methods that it may
+    need to override in case of reuse 'CCompilerOpt' in environment
+    different than of what NumPy has.
+
+    Parameters
+    ----------
+    ccompiler : `CCompiler`
+        The generate instance that returned from `distutils.ccompiler.new_compiler()`.
+    """
+    def __init__(self, ccompiler):
+        self._ccompiler = ccompiler
+
+    def dist_compile(self, sources, flags, **kwargs):
+        """Wrap CCompiler.compile()"""
```

Comment:
last discovered one hopefully.
we have another issue here with parallel compiling, each dispatched target requires variant flags so I had to call `compile()` multiple times since it doesn't provide an argument to specify flags for each source.

I have two ways to take in order to fix this issue :

- by patching the custom `compile()` that defined in `ccompiler.py` to allow non-block calls
   through callable argument.
- by handel parallel compiling internally though threads in here.

@mattip, @rgommers, any ideas?

Reply:
/cc, @juliantaylor

Reply:
> we have another issue here with parallel compiling, each dispatched target requires variant flags so I had to call `compile()` multiple times since it doesn't provide an argument to specify flags for each source.

No way around that indeed.



>     * by patching the custom `compile()` that defined in `ccompiler.py` to allow non-block calls
>       through callable argument.

Hmm that doesn't sound attractive, I wouldn't want to monkeypatch `distutils` for some potential speed gain.

> by handel parallel compiling internally though threads in here.

Doesn't seem worth it either to be honest. We recently detected a bug in the Fortran support for `--parallel`, that stuff is a bit painful to get right. Let's not worry about compile speed of this for now - first get the patch in, worry about possible compile speedups later.

Reply:
I feel it worth it in the near future since we will have a lot of dispatch-able sources, but it's okay for now.